cuda/matrix/csr_kernels.cu

```diff
@@ -246,33 +247,158 @@ void classical_spmv(syn::value_list<int, subwarp_size>,
+                nwarps, static_cast<IndexType>(a->get_size()[0]),
+                as_cuda_type(a->get_const_values()), a->get_const_col_idxs(),
+                as_cuda_type(a->get_const_row_ptrs()),
+                as_cuda_type(a->get_const_srow()),
+                as_cuda_type(b->get_const_values()),
+                as_cuda_type(b->get_stride()), as_cuda_type(c->get_values()),
+                as_cuda_type(c->get_stride()));
         }
+    } else {
+        GKO_NOT_SUPPORTED(nwarps);
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+bool try_general_sparselib_spmv(std::shared_ptr<const CudaExecutor> exec,
+                                const ValueType* alpha,
+                                const matrix::Csr<ValueType, IndexType>* a,
+                                const matrix::Dense<ValueType>* b,
+                                const ValueType* beta,
+                                matrix::Dense<ValueType>* c)
+{
+    auto handle = exec->get_cusparse_handle();
+#if defined(CUDA_VERSION) && (CUDA_VERSION < 11000)
+    if (b->get_stride() != 1 || c->get_stride() != 1) {
+        return false;
+    }
+
+    auto descr = cusparse::create_mat_descr();
+    auto row_ptrs = a->get_const_row_ptrs();
+    auto col_idxs = a->get_const_col_idxs();
+    cusparse::spmv(handle, CUSPARSE_OPERATION_NON_TRANSPOSE, a->get_size()[0],
+                   a->get_size()[1], a->get_num_stored_elements(), alpha, descr,
+                   a->get_const_values(), row_ptrs, col_idxs,
+                   b->get_const_values(), beta, c->get_values());
+
+    cusparse::destroy(descr);
+#else  // CUDA_VERSION >= 11000
+    cusparseOperation_t trans = CUSPARSE_OPERATION_NON_TRANSPOSE;
+    auto row_ptrs = const_cast<IndexType*>(a->get_const_row_ptrs());
+    auto col_idxs = const_cast<IndexType*>(a->get_const_col_idxs());
+    auto values = const_cast<ValueType*>(a->get_const_values());
+    auto mat = cusparse::create_csr(a->get_size()[0], a->get_size()[1],
+                                    a->get_num_stored_elements(), row_ptrs,
+                                    col_idxs, values);
+    auto b_val = const_cast<ValueType*>(b->get_const_values());
+    auto c_val = c->get_values();
+    if (b->get_stride() == 1 && c->get_stride() == 1) {
+        auto vecb = cusparse::create_dnvec(b->get_size()[0], b_val);
+        auto vecc = cusparse::create_dnvec(c->get_size()[0], c_val);
+        cusparseSpMVAlg_t alg = CUSPARSE_CSRMV_ALG1;
+        size_type buffer_size = 0;
+        cusparse::spmv_buffersize<ValueType>(handle, trans, alpha, mat, vecb,
+                                             beta, vecc, alg, &buffer_size);
+
+        gko::Array<char> buffer_array(exec, buffer_size);
+        auto buffer = buffer_array.get_data();
+        cusparse::spmv<ValueType>(handle, trans, alpha, mat, vecb, beta, vecc,
+                                  alg, buffer);
+        cusparse::destroy(vecb);
+        cusparse::destroy(vecc);
+    } else {
```

Comment:
Just a note : For multiple apply like in a solver kernel, it makes sense to split this into two steps. A prepare and an apply step, so that we can create these objects in a prepare, maintain a state of the apply in some way and destroy it afterwards. This will definitely improve performance of sparselib functionality. 

I wouldn't necessarily do this here but probably in another PR.

Reply:
Yes, I think the right place to store the buffer would be a generalized `srow`, related to #320.

Reply:
Yes, I agree. I guess we can tackle this in another PR. I have noted this in #320.